_posts/2021/2021-01-25-memoize-expensive-code.md

```diff
@@ -0,0 +1,75 @@
+---
+title: "Memoize Expensive Code"
+description: ""
+layout: article
+category: ruby
+image:
+  base: '2021/memoize-expensive-code'
+  alt: "Red Floppy Disk"
+  credit:  Fredy Jacob
+  source: "https://unsplash.com/photos/t0SlmanfFcg"
+---
+
+Memoization is a performance optimization where the result of a slow or non-performant piece of code is stored and when the expensive code is called again, the stored value is returned.
+
+It is a specific form of caching that is relatively elegant to read and implement in Ruby.
+
+
+## Instead of ...
+
+...repeating potentially expensive calculations:
+
+```ruby
+class OldTimeySweetShop
+
+  def average_sweets_per_jar
+    sweet_count / glass_jars.count
+  end
+
+  def sweet_count
+    glass_jars.sum do |jar|
+      jar.count_each_sweet_by_hand
+    end
+  end
+end
+```
+
+
+## Instead...
+
+```ruby
+class OldTimeySweetShop
+`
+  # leave the same
+
+  def sweet_count
+    @sweet_count ||= glass_jars.sum do |jar|
+      jar.count_each_sweet_by_hand
+    end
+  end
+end
+
+shop = OldTimeySweetShop.new
+shop.sweet_count
+#=> 4000003
+shop.sweet_count
+#=> 4000003 # this one will be super quick!
+```
+
+
+## Why?
+
+Primarily this is a performance optimization. The first call of the method stores the result of an expensive operation in an instance variable that is then returned.
+
+When the method is called again the stored instance value is used and the expensive calculation is avoided.
+
+In typical Rails applications, I've found this patten to be most useful to optimize expensive database calls and to temporarily cache API requests to external services.
+
+
+## Why not?
+
+A piece of code can only be memoized _only if_ calling the code again would have the same output as replacing that function call with its return value.
```

Comment:
You repeat 'only' twice in the first sentence. I think you can remove the first one.

Reply:
```suggestion
A piece of code can be memoized _only if_ calling the code again would have the same output as replacing that function call with its return value.
```